File: Programmers/[49995] 쿠키 구입/n__aj22/49995.py
def solution(cookie):
    answer = -1

    sum_list = []
    for i in range(len(cookie)+1):
        sum_list.append(sum(cookie[0:i]))

    half_sum = sum_list[-1]//2

    for m in range(len(cookie)):
        for l in range(m+1):
            #l~m
            now_sum = sum_list[m+1]-sum_list[l]
            last_sum = sum_list[len(cookie)] - sum_list[m+1]
            if now_sum>last_sum: # 남은 쿠키 개수보다 현재 확인할 합계가 작으면 확인할 필요 없음
                continue
            # if now_sum>half_sum: #현재 확인할 합계가 전체 개수의 반틈보다 작으면 확인할 필요 없음
            #     continue
            if answer!=-1 and now_sum<answer: #현재 확인할 합계가 현재 찾은 답보다 작으면 확인할 필요 없음
                continue
            for r in range(m+1, len(cookie)):
                #m+1 ~ r
                check_sum = sum_list[r+1] - sum_list[m+1]
                if(check_sum == now_sum):
                    answer = max(answer, now_sum)
                    break
                elif(check_sum>now_sum):
                    break
    if answer == -1:
        answer = 0
    return answer

if __name__ == "__main__":
    cookie = [[1,1,2,3], [1,2,4,5]]

    for c in cookie:
        print(solution(c))

File: Programmers/[49995] 쿠키 구입/lululalyang/_49995.java
public class _49995 {

	public static void main(String[] args){
		int[] cookie = {1,2,4,5};
		System.out.println(solution(cookie));
	}
	
	private static int solution(int[] cookie) {
		int MAX = 0;
		int N = cookie.length;
		int[] sum = new int[N+1];
		for(int i=1; i<=N; i++) {
			sum[i] = sum[i-1] + cookie[i-1];
		}
		
		for(int m=0; m<N; m++) {
			for(int l=1; l<m+1; l++) {
				int bro1 = sum[m] - sum[l-1]; // l ~ m
				int remain = sum[N] - sum[m]; // 남은 개수 (bro2가 가질 수 있는 최대 개수)
				
				if(bro1 > remain)	continue; // bro2 최대 개수가 bro1 보다 작으면 불가능 => pass
				if(MAX > bro1)	continue; // 이미 구한 답보다 bro1이 작다면 확인할 필요 없다 => pass
				
				for(int r=m+1; r<=N; r++) {
					int bro2 = sum[r] - sum[m]; // (m+1) ~ r
					if(bro1 == bro2) {
						MAX = Math.max(bro2, MAX);
						break;
					}else if(bro1 < bro2)	break;
				}
			}
		}
		
		return MAX;
	}	
}
